### ISSUE TITLE
Warn for `set -u` variables that will fail to expand

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#2779]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
- My shellcheck version (`shellcheck --version` or "online"): 
- [ ] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
set -Eeuo pipefail

if [ -z "${var}" ] ; then
....
```

#### Here's what shellcheck currently says:

_Nothing_

#### Here's what I wanted or expected to see:

`var` is not defined, and you have `set -u` activated. That will result in failure.

I know that a "quick run" would result in uncovering that issue.
Waiting to test that "in a complex CI scenario" might be unwanted.

**Comment by aeiplatform** [comment]
Shellcheck would need to know env etc.

**Comment by felipecrs** [comment]
> Shellcheck would need to know env etc.

I honestly don't think so, can you elaborate? User is not talking about environment variables (conventionally UPPER_CASE). For that we have `# shellcheck enable=check-unassigned-uppercase`.


I believe [SC2154](https://www.shellcheck.net/wiki/SC2154) is a check that already covers this use case (i.e. as long as the user fixes SC2154, the variable will not fail to expand).

But I believe there is a problem with this check, which is triggered by the example provided:

https://github.com/koalaman/shellcheck/assets/29582865/57a39600-8b67-4246-9f7e-01f5113cecc9

**Comment by felipecrs** [comment]
But I totally agree with the suggestion, this would be an extremely helpful check which happens very frequently to my colleagues which are enforced to use `set -u` but are not used to it.

**Comment by aeiplatform** [comment]
This suggestion is like duplicated all over this place https://github.com/koalaman/shellcheck/issues/2771

**Comment by aeiplatform** [comment]
It's only convention, not an axiom.
e.g. `http_proxy`

**Comment by felipecrs** [comment]
But this suggestion has some uniqueness in it, which is the mention of `set -u`. Without `set -u`, the script doesn't stop and potentially works, with `set -u` the script will fail no matter what, so it is indeed less tolerable.

**Comment by aeiplatform** [comment]
The uniqueness that you are referencing, does not matter, because shellcheck is static analysis tool not a shell interpreter.

**Comment by wileyhy** [comment]
Such a warning message would duplicate an already existing and available one.

**Comment by felipecrs** [comment]
> Such a warning message would duplicate an already existing and available one.

Not for the use case demonstrated above, where no warning is shown at all. So, either the existing rule should be extended or a new one should be added to cover this situation.

```bash
#!/bin/bash

set -u

# No warning is shown for the line below
if [[ -n "${var}" ]]; then
  true
fi
```

**Comment by wileyhy** [comment]
It works on my machine. Which version of bash are you using for this?

[liveuser@localhost-live shellcheck_2779]$ bash --version
GNU bash, version 5.2.15(1)-release (x86_64-redhat-linux-gnu)

########
[liveuser@localhost-live shellcheck_2779]$ cat ./2779.bash
#!/bin/bash -x
set -u
if [[ -n "${var}" ]]; then
  true
fi

[liveuser@localhost-live shellcheck_2779]$ ./2779.bash
+ set -u
./2779.bash: line 3: var: unbound variable
########

**Comment by wileyhy** [comment]
Also...

[liveuser@localhost-live shellcheck_2779]$ cat ./2779-b.bash
#!/bin/sh -x
set -eu
if [ -z "${var}" ]
then 
  :
fi

[liveuser@localhost-live shellcheck_2779]$ ./2779-b.bash
+ set -eu
./2779-b.bash: line 3: var: unbound variable

**Comment by felipecrs** [comment]
Of course the issue is caught from Bash. But we are talking about ShellCheck, which should catch the issue to prevent it from happening in Bash.

**Comment by TinCanTech** [comment]
This first:
```
# No warning is shown for the line below
if [[ -n "${var}" ]]; then
  true
fi
```
My guess is that `shellcheck` is trying to be less noisy by not reporting this undeclared variable `var`.  The script is deliberately testing `var`, so there is no need for `shellcheck` to report it.

Then add `set -u`, which does cause the script to fail in `bash`.

The problem appears to be that `shellcheck` is not taking `set -u` into account.

Does `shellcheck` ever account for any shell options ?

If `shellcheck` does account for shell options then, _I would tend to agree_, that it should take `-u` into account, in the manner expressed in this issue.

**Comment by felipecrs** [comment]
I believe shellcheck takes set -e into account for some checks.

**Comment by TinCanTech** [comment]
`nounset` is far less _troublesome_ than `errexit`.

Because your example is specifically testing the value of `var`, I think `shellcheck` takes that as an OK-to-ignore signal.  But, I do agree, with `nounset`, your example is fatal.  So, _why not report it ?_

This example goes a little further:
```
#!/bin/bash

set -u

if [[ -n "${foo}" ]]; then
	echo "${foo}"
else
	echo "${bar}"
fi
```
```
In example line 9:
	echo "${bar}"
              ^----^ SC2154 (warning): bar is referenced but not assigned.
```
I think it may simply be an over-sight in `shellcheck`, which can be corrected.

**Comment by TinCanTech** [comment]
There is another way to look at this problem:
```
#!/bin/bash

set -u

# No warning is shown for the line below
if [[ -n "${var}" ]]; then
  true
fi
```
This is basically bugged code, because of `-u`, which needs to be fixed

**Comment by felipecrs** [comment]
> There is another way to look at this problem:
> ```
> #!/bin/bash
> 
> set -u
> 
> # No warning is shown for the line below
> if [[ -n "${var}" ]]; then
>   true
> fi
> ```
> This is basically bugged code, because of `-u`, which needs to be fixed

Yes, which I think shellcheck should help detect.

**Comment by Felixoid** [comment]
Here's another place where I counted on `shellcheck`, but my buggy code passed without complains:

```
> cat broken-shellcheck.sh
#!/bin/bash

set -euo pipefail

DATA_DIR="/"
# checking $DATA_DIR for initialization
if [ -d "${DATA_DIR%/}/non-existing-data" ]; then
    DATABASE_ALREADY_EXISTS='true'
fi

echo "$DATABASE_ALREADY_EXISTS"
> shellcheck -o check-unassigned-uppercase broken-shellcheck.sh
> bash -x broken-shellcheck.sh
+ set -euo pipefail
+ DATA_DIR=/
+ '[' -d /non-existing-data ']'
broken-shellcheck.sh: line 11: DATABASE_ALREADY_EXISTS: unbound variable
```
